ompi/communicator/comm.c

```diff
@@ -1010,130 +1143,73 @@ int ompi_comm_split_type (ompi_communicator_t *comm, int split_type, int key,
-     */
-    if (MPI_COMM_TYPE_HW_UNGUIDED == global_orig_split_type) {
-        *newcomm = MPI_COMM_NULL;
-        return OMPI_SUCCESS;
-    }
-
-    /* Step 2: Build potential communicator groups. If any ranks will not be part of
-     * the ultimate communicator we will drop them later. This saves doing an extra
-     * allgather on the whole communicator. By using ompi_comm_split() later only
-     * if needed we 1) optimized the common case (no MPI_UNDEFINED and no reorder),
-     * and 2) limit the allgather to a smaller set of peers in the uncommon case. */
-    /* --------------------------------------------------------- */
-
-    /* allowed splitting types:
-       CLUSTER
-       CU
-       HOST
-       BOARD
-       NODE
-       NUMA
-       SOCKET
-       L3CACHE
-       L2CACHE
-       L1CACHE
-       CORE
-       HWTHREAD
-       Even though HWTHREAD/CORE etc. is overkill they are here for consistency.
-       They will most likely return a communicator which is equal to MPI_COMM_SELF
-       Unless oversubscribing.
-    */
-
-    /* how many ranks are potentially participating and on my node? */
-    rc = ompi_comm_split_type_get_part (comm->c_local_group, global_split_type, &lranks, &my_size);
-    if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-        return rc;
-    }
-
-    /* Step 3: determine all the information for the remote group */
-    /* --------------------------------------------------------- */
-    if (inter) {
-        rc = ompi_comm_split_type_get_part (comm->c_remote_group, global_split_type, &rranks, &my_rsize);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            free (lranks);
-            return rc;
+    /* MPI_COMM_TYPE_HW_GUIDED: Check if 'value' the same at all ranks */
+    if (tmp[4] != -tmp[5]) {
+        if (0 == ompi_comm_rank(comm)) {
+            opal_output(0, "Error: Mismatched info values for MPI_COMM_TYPE_HW_GUIDED");
         }
+        return OMPI_ERR_BAD_PARAM;
     }
 
-    /* set the CID allgather mode to the appropriate one for the communicator */
-    mode = inter ? OMPI_COMM_CID_INTER : OMPI_COMM_CID_INTRA;
-
-    /* Step 4: set up the communicator                           */
-    /* --------------------------------------------------------- */
-    /* Create the communicator finally */
-
-    do {
-        rc = ompi_comm_set (&newcomp, comm, my_size, lranks, my_rsize,
-                            rranks, NULL, comm->error_handler, NULL, NULL, 0);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            break;
-        }
-
-        /* Determine context id. It is identical to f_2_c_handle */
-        rc = ompi_comm_nextcid (newcomp, comm, NULL, NULL, NULL, false, mode);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            break;
-        }
+    if (MPI_COMM_TYPE_HW_UNGUIDED == global_orig_split_type) {
+        /* Handle MPI_COMM_TYPE_HW_UNGUIDED communicator split. */
+        return ompi_comm_split_unguided( comm, info, newcomm );
```

Comment:
Pass down some of the state computation variables so you can call `ompi_comm_split_type_core` directly from `ompi_comm_split_unguided`
```suggestion
        return ompi_comm_split_unguided( comm, key, info, need_split, no_reorder, no_undefined, newcomm );
```